### PR TITLE
Feat: Partial implementation: Status & RegionSkill. ✔

### PR DESCRIPTION
**Pull request by Sparkymod**
### SkillMetadata
- Parse more data from SkillAdditionalData.
- Parsing missing skills, from additionalEffect.
- Adding type SkillAttack, to handle ConditionsSkills and more data to parse.

### Skill Cast
- 'Temporary fix' for some methods that were giving errors, when finding the correct property for the skill.
- Implementing SkillAdditionalData info to then use.
- Adding 'IsHeal` check for skills that meet the requirements.

### Status
- Adding enum of SkillTypes. (if is buff, status, debuff...)
- Implement a StatusHandler type to handle the status calls.

### SkillHandler
- New mode: 0x02 - Casting continuously skill.
- Moving the StatusCheck from Player.Cast to SkillHandler.
- Now players can heal or damage, depends on the skill. (Partially apply).
- 

### Region Effects
- Partially Implement. When casting skills, some others skills are called from the first one.
- Implement a RegionSkillsHandler type to handle the effects calls.

### Buff
- Changing `Owner` to `Target`, since it was confusing, and is a better approach to what it does.

Open to suggestions 😊

### CODE REVIEW

File: GameDataParser/Parsers/SkillParser.cs [hunk 1]
```diff
@@ -120,27 +135,53 @@ protected override List<SkillMetadata> Parse()
             foreach (PackFileEntry entry in Resources.XmlFiles)
             {
                 if (!entry.Name.StartsWith("additionaleffect"))
-                {
-                    continue;
-                }
+                { continue; }
```
Comment (AlanMorel) [inline GameDataParser/Parsers/SkillParser.cs:138]: Format?

File: GameDataParser/Parsers/SkillParser.cs [hunk 2]
```diff
@@ -120,27 +135,53 @@ protected override List<SkillMetadata> Parse()
...
                 XmlNodeList levels = document.SelectNodes("/ms2/level");
-
                 int skillId = int.Parse(Path.GetFileNameWithoutExtension(entry.Name));
+
                 if (skillList.Select(x => x.SkillId).Contains(skillId))
                 {
                     foreach (XmlNode level in levels)
                     {
-                        int duration = int.Parse(level.SelectSingleNode("BasicProperty").Attributes["durationTick"]?.Value ?? "0");
+                        SkillAdditionalData data = ParseSkillData(level);
```
Comment (AlanMorel) [inline GameDataParser/Parsers/SkillParser.cs:148]: Why not only define this only when actually necessary?

File: MapleServer2/PacketHandlers/Game/SkillHandler.cs [hunk 3]
```diff
@@ -171,19 +181,51 @@ private static void HandleAoeDamage(GameSession session, PacketReader packet)
...
             byte mode = packet.ReadByte();
-            int unk1 = packet.ReadInt();
-            int unk2 = packet.ReadInt();
+            int unknown = packet.ReadInt();
+            int unknown2 = packet.ReadInt();
             CoordF coord = packet.Read<CoordF>();
             CoordF coord1 = packet.Read<CoordF>();
+
+            // TODO: Verify rest of skills to proc correctly.
+            if (SkillUsePacket.SkillCastMap[skillSN].GetConditionSkill() != null)
```
Comment (AlanMorel) [inline MapleServer2/PacketHandlers/Game/SkillHandler.cs:221]: Invert and return instead

File: MapleServer2/Types/SkillCast.cs [hunk 4]
```diff
@@ -42,22 +44,50 @@ public SkillCast(int id, short level, long skillSN, int unkValue)
...
+        public int MaxStack() => SkillMetadataStorage.GetSkill(SkillId)?.SkillLevels.Find(s => s.Level == SkillLevel).SkillAdditionalData.MaxStack ?? 1;
+
+        public IEnumerable<int> GetConditionSkill() => SkillMetadataStorage.GetSkill(SkillId)?.SkillLevels.Find(s => s.Level == SkillLevel).SkillAttacks.ConditionSkillIds ?? null;
 
-        public int DurationTick() => SkillMetadataStorage.GetSkill(SkillId).SkillLevels.Find(s => s.Level == SkillLevel).SkillAdditionalData.Duration;
+        public bool IsHeal() => VerifySkillTypeOf(SkillType.None, SkillSubType.Status, BuffType.Buff, BuffSubType.Recovery);
+
+        private bool VerifySkillTypeOf(SkillType type, SkillSubType subType, BuffType buffType, BuffSubType buffSubType)
+        {
+            Maple2Storage.Types.Metadata.SkillMetadata skillData = SkillMetadataStorage.GetSkill(SkillId);
```
Comment (AlanMorel) [inline MapleServer2/Types/SkillCast.cs:81]: Why not import SkillMetadata?

**Review by AlanMorel — CHANGES_REQUESTED** [review]

File: MapleServer2/Types/SkillCast.cs [hunk 5]
```diff
@@ -42,22 +46,66 @@ public SkillCast(int id, short level, long skillSN, int unkValue)
...
+        public DamageTypeId GetSkillDamageType()
+        {
+            if (SkillMetadataStorage.GetSkill(SkillId) != null)
+            {
+                return (DamageTypeId) SkillMetadataStorage.GetSkill(SkillId).DamageType;
+            }
+            else
+            {
+                return DamageTypeId.None;
+            }
```
Comment (AngeloTadeucci) [inline MapleServer2/Types/SkillCast.cs:66]: Maybe it's too dank...

```suggestion
            return SkillMetadataStorage.GetSkill(SkillId) != null ? (DamageTypeId) SkillMetadataStorage.GetSkill(SkillId).DamageType : DamageTypeId.None;
```

**Review by AngeloTadeucci — APPROVED** [review]
LGTM!

File: MapleServer2/Types/SkillCast.cs (hunk 5 above)
Comment (AngeloTadeucci) [reply MapleServer2/Types/SkillCast.cs:66]: Even more dank

`        public DamageTypeId GetSkillDamageType() => SkillMetadataStorage.GetSkill(SkillId) != null ? (DamageTypeId) SkillMetadataStorage.GetSkill(SkillId).DamageType : DamageTypeId.None;
`

Comment (Sparkymod) [reply MapleServer2/Types/SkillCast.cs:68]: i recently add `GetSkillMetadata()` to reduce the calls and more readable, suggest the change with it so i can commit it

File: MapleServer2/Types/SkillCast.cs [hunk 6]
```diff
@@ -42,22 +46,66 @@ public SkillCast(int id, short level, long skillSN, int unkValue)
...
+        }
```
Comment (AngeloTadeucci) [inline MapleServer2/Types/SkillCast.cs:69]:
```suggestion
        public DamageTypeId GetSkillDamageType() => (DamageTypeId) (GetSkillMetadata()?.DamageType ?? 0);
```

File: MapleServer2/Types/SkillCast.cs [hunk 7]
```diff
@@ -42,22 +46,96 @@ public SkillCast(int id, short level, long skillSN, int unkValue)
...
+            if (VerifySkillTypeOf(SkillType.None, SkillSubType.Status, BuffType.Debuff, BuffSubType.Entity))
+            {
+                return true;
+            }
+            else if (VerifySkillTypeOf(BuffType.Debuff, BuffSubType.Entity))
+            {
+                return true;
+            }
+            return false;
+        }
```
Comment (AngeloTadeucci) [inline MapleServer2/Types/SkillCast.cs:98]:
```suggestion
                public bool IsDebuffToEntity() => VerifySkillTypeOf(SkillType.None, SkillSubType.Status, BuffType.Debuff, BuffSubType.Entity) || VerifySkillTypeOf(BuffType.Debuff, BuffSubType.Entity);

```

**Review by AngeloTadeucci — APPROVED** [review]

**Pull request merged**